src/sensors/elec.rs

```diff
@@ -14,21 +26,142 @@ pub fn filter_pde(pde: f32) -> bool {
     random::<f32>() < pde
 }
 
+pub fn gaussian_sampler(mu: f32, sigma: f32) -> Box<dyn Fn() -> f32> {
+    let dist = Normal::new(mu, sigma).unwrap();
+    Box::new(move || dist.sample(&mut rand::thread_rng()))
+}
+
+// Could this be a FROM? How to use the Group here to avoid too many collects?
+fn qt_min_time(event_id: u32, sensor_id: u32, hits: Vec<MCSensorHit>) -> MCQT {
+    let first_time = hits.iter().fold(f32::INFINITY, |min_val, val| val.time.min(min_val));
+    MCQT { event_id, sensor_id, q: hits.len() as u32, t: ns(first_time) }
+}
+
+/// combine the individual hits into information for each sensor.
+/// Currently uses the minimum as the time for each sensor which
+/// might prove too inflexible. Should be generalised.
+pub fn combine_sensor_hits(event_id: u32, hits: Vec<MCSensorHit>) -> Vec<MCQT> {
+    hits.into_iter()
+        .group_by(|h| h.sensor_id)
+        .into_iter()
+        .map(|(s, grp)| qt_min_time(event_id, s, grp.collect()))
+        .collect()
+}
+
+// This seems like something that would be in a crate somewhere.
+fn dot((x1,y1,z1): (Length, Length, Length), (x2,y2,z2): (Length, Length, Length)) -> Area {
+    x1*x2 + y1*y2 + z1*z2
+}
+
+// Dot product separation for now, need to implement other algorithm(s)
+fn dot_product_clusters(hits: &[MCQT], xyzs: &SensorMap, threshold: f32, min_sensors: usize) -> Option<(Vec<MCQT>, Vec<MCQT>)> {
+    let max_sens = hits.iter().max_by_key(|e| e.q).map(|e| e.sensor_id)?;
+    let &(xm, ym, zm) = xyzs.get(&max_sens)?;
+    let (c1, c2): (Vec<MCQT>, Vec<MCQT>) = hits.iter()
+            .filter(|&MCQT { q, .. }| *q as f32 > threshold)
+            .partition(|&MCQT { sensor_id, .. }| dot((xm, ym, zm), *xyzs.get(sensor_id).unwrap()) > Area::ZERO);
+    if c1.len() >= min_sensors && c2.len() >= min_sensors { Some((c1, c2)) } else { None }
+}
+
+/// Reconstruct LORs based on the barycentre of the 2 selected clusters (dot prod) without DOI.
+fn lor_from_sensor_positions(hits: &[MCQT], xyzs: &SensorMap, threshold: f32, min_sensors: usize) -> Option<Hdf5Lor> {
+    let (c1, c2) = dot_product_clusters(hits, xyzs, threshold, min_sensors)?;
+    // TODO time just taken as minimum of minima. Averaging used elsewhere.
+    let b1 = sipm_charge_barycentre(&c1, xyzs)?;
+    let b2 = sipm_charge_barycentre(&c2, xyzs)?;
+    let q1 = c1.iter().map(|&MCQT { q, .. }| q as f32).sum();
+    let q2 = c2.iter().map(|&MCQT { q, .. }| q as f32).sum();
+    // Just dummy time for now until we work out the best way to sort.
+    Some(Hdf5Lor {
+        dt: ns_(b2.3 - b1.3),
+        x1: mm_(b1.0), y1: mm_(b1.1), z1: mm_(b1.2),
+        x2: mm_(b2.0), y2: mm_(b2.1), z2: mm_(b2.2),
+        q1, q2, E1: f32::NAN, E2: f32::NAN,
+    })
```

Comment:
Aside: it bothers me that this interface requires the human to recall that the units in our `Hdf5Lor` data are `ns` and `mm`. This is asking for bugs. OK, the `ns_` and `mm_` are in-your-face, so anyone copying examples is less *likely* to make a mistake, but still. (In fact, I made a mistake and typed `ps` instead of `ns` while writing this comment. Had I made that mistake in the code, the compiler would not have caught it. Well, maybe, because the wrong converter might not have been imported). We really need a constructor which accepts `Time` and `Length` and makes the correct choice behind the interface, as a sole point of conversion.

But the struct initialization syntax does give us something very much like keyword arguments, and that has its value too.